### DISCUSSION TITLE
Robyn Cluster Questions

### DISCUSSION THREAD

**Discussion by Runsicker11** [facebookexperimental/Robyn#325] [General]
I really like the idea of the clusters, reducing the number of models to choose from significantly, but I want to understand what is going on. How does Robyn determine which model to keep? 

Are they clustering the results into new models or are they clustering into groups and then selecting the best model from each group?

Also can the user determine how many clusters to create. It seems the default is 6.

Thanks

**Comment by laresbernardo** [comment]
Hi @Runsicker11 

We are about to update the documentation on this matter soon, but I'll briefly explain what `robyn_clusters()` do. The idea behind it is to be able to reduce the number of models to compare/select from, by picking the "best most different models" within Pareto front results. This is how we do it:

1. Using the pool of models on Pareto fronts obtained, we run k-means clustering using (normalized) ROI data on each (paid) media variable
2. When `k = "auto"` (which is the default), we calculate the WSS (Within Group Sum of Squares) on k-means clustering using k = 1 to k = 20 to find the "best automatic k". We pick the one k whose WSS reduction doesn't change more than 5% from the previous k. You could also override this by setting k manually (on `robyn_run(..., cluster = TRUE)`, `robyn_outputs(..., clusters = TRUE)`, or `robyn_clusters()`).
3. After we have run k-means on all Pareto front models, using the defined k, we pick those "best models" that have the lowest normalized combined errors (NRMSE, DECOM.RSSD, and MAPE if calibrated was used). 

When you run `robyn_clusters()` you'll get a list of results: the data used to calculate the clusters with each model's cluster-ID, and some visualizations on WSS-k selection, ROI per media on winner models, and correlations of ROI by cluster to understand each cluster's content models better.

I'm attaching a couple of screenshots of these plots using dummy data and the output CSV.

![pareto_clusters_detail](https://user-images.githubusercontent.com/25909073/155625010-89e9480e-9642-4149-94d1-6349dcda6e01.png)
![pareto_clusters_wss](https://user-images.githubusercontent.com/25909073/155625016-db156913-fdcb-4c4b-b776-71d3ecad0ab8.png)
[pareto_clusters.csv](https://github.com/facebookexperimental/Robyn/files/8137224/pareto_clusters.csv)
